std/format/internal/write.d

```diff
@@ -453,194 +435,65 @@ if (is(FloatingPointTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
-                            fs.precision == fs.UNSPECIFIED ? -1 : fs.precision,
-                            tval);
-        }();
-
-        enforceFmt(n >= 0, "floating point formatting failure");
-
-        len = min(n, buf2.length-1);
-        buf = buf2;
-
-        if (fs.flSeparator && !inf && !nan)
-        {
-            ptrdiff_t indexOfRemovable()
-            {
-                if (len < 2)
-                    return -1;
+        import std.math.traits : isInfinity;
+        import std.math.operations : nextUp, nextDown;
 
-                size_t start = (buf[0 .. 1].indexOfAny(" 0123456789") == -1) ? 1 : 0;
-                if (len < 2 + start)
-                    return -1;
-                if ((buf[start] == ' ') || (buf[start] == '0' && buf[start + 1] != '.'))
-                    return start;
+        // reals that are not supported by printFloat are cast to double.
+        double tval = val;
 
-                return -1;
-            }
-
-            ptrdiff_t dot, firstDigit, ePos, dotIdx, firstLen;
-            size_t separatorScoreCnt;
-
-            while (true)
-            {
-                dot = buf[0 .. len].indexOf('.');
-                firstDigit = buf[0 .. len].indexOfAny("0123456789");
-                ePos = buf[0 .. len].indexOf('e');
-                dotIdx = dot == -1 ? ePos == -1 ? len : ePos : dot;
-
-                firstLen = dotIdx - firstDigit;
-                separatorScoreCnt = (firstLen > 0) ? (firstLen - 1) / fs.separators : 0;
-
-                ptrdiff_t removableIdx = (len + separatorScoreCnt > fs.width) ? indexOfRemovable() : -1;
-                if ((removableIdx != -1) &&
-                    ((firstLen - (buf[removableIdx] == '0' ? 2 : 1)) / fs.separators + len - 1 >= fs.width))
-                {
-                    buf[removableIdx .. $ - 1] = buf.dup[removableIdx + 1 .. $];
-                    len--;
-                }
-                else
-                    break;
-            }
-
-            immutable afterDotIdx = (ePos != -1) ? ePos : len;
+        // Numbers greater than double.max are converted to double.max:
+        if (val > double.max && !isInfinity(val))
+            tval = double.max;
+        if (val < -double.max && !isInfinity(val))
+            tval = -double.max;
 
-            // plus, minus, prefix
-            if (firstDigit > 0)
-            {
-                put(w, buf[0 .. firstDigit]);
-            }
-
-            // digits until dot with separator
-            for (auto j = 0; j < firstLen; ++j)
-            {
-                if (j > 0 && (firstLen - j) % fs.separators == 0)
-                {
-                    put(w, fs.separatorChar);
-                }
-                put(w, buf[j + firstDigit]);
-            }
+        // Numbers between the smallest representable double subnormal and 0.0
+        // are converted to the smallest representable double subnormal:
+        if (val > 0 && val < nextUp(0.0))
+            tval = nextUp(0.0);
```

Comment:
`nextUp(0.0)` and `nextDown(-0.0)` can be compile time constants. e.g.
```
enum doubleLowest = nextUp(0.0);
enum doubleNegativeLowest = nextDown(-0.0);
```

This ensures no run-time calls. It also makes sense from the perspective that these are essentially derived from the  IEEE standard, so they have fixed values. Could also specify the type for clarity. e.g. `enum double xyz`. Not sure about the Phobos guideline on this.